### DISCUSSION TITLE
Documentation on setting up external authentication

### DISCUSSION THREAD

**Discussion by RonRademaker** [healthchecks/healthchecks#743] [Q&A] [answered]
Hi all,

I'm looking for some documentation on setting up external authentication with the REMOTE_USER_HEADER feature. Is there something available somewhere, I'm thinking maybe a blog describing step by step how to set this up or some example configurations to get me started.

What I'm trying to do is connect healthchecks to my existing AWS Cognito setup somehow. Users in the the cognito user pool with a specific group should be granted access. 

Thanks!
Ron

**Comment by cuu508** [comment, marked as answer]
Hi Ron,

I'm assuming you've already seen https://healthchecks.io/docs/self_hosted_configuration/#REMOTE_USER_HEADER?

I haven't seen examples of integrating with AWS Cognito, but perhaps others will chime in.

**Reply by RonRademaker** [reply]
Hi, yes I read that part but I think I need a bit more to get started (after reading it is couple of times I still don't know where to start :) ). Some examples on how to actually use it and set it up would be nice, not necessarily specifically for AWS cognito.

**Reply by cuu508** [reply]
As a rough example:

* Set `REMOTE_USER_HEADER=HTTP_X_USER`
* Run Healthchecks behind a reverse proxy. Configure the reverse proxy to add a `X-User: someusername@example.org` request header to a request if the user should be authenticated as `someusername@example.org`. 
* Important: make sure the reverse proxy does not pass on a `X-User` request header sent by site visitor. Without this, any visitor could pose as any user, just by setting a request header.

Healthchecks will automatically log in the user specified in the `X-User` request header. If the user does not exist, it will create it first.

The specific header value does not matter, you could use `HTTP_COGNITO_USER` (which would correspond to `Cognito-User` request header), or any other value. 

The missing bit in this example is, how to configure the reverse proxy to integrate with AWS Cognito. Unfortunately I cannot help with that part.

**Reply by RonRademaker** [reply]
Thanks, looks like I'm getting somewhere now 🎉 

Which does bring me to a next hurdle, users are not usually already known in healthchecks and it seems their account is created automatically (which is nice). However, how do I automatically grant them access to the correct project?

**Reply by cuu508** [reply]
There is no built-in functionality for automatically granting access to a project. Do you have a lot of users, or does the set of users needing access change often?

**Reply by RonRademaker** [reply]
My issue is not with the amount of users or changes, but more with the amount of applications accessible by the users. I'm sure we will forget adding the users manually to healthchecks (in the landscape, it is one of the applications users will use the least and is easily forgotten). 

Another approach would also be okay for me, for example using an API call to create and assign a user (but I don't see any user management API endpoints in the docs).

**Reply by RonRademaker** [reply]
For anyone reading this in the future, the following configuration for the oauth2-proxy works for me (development setup, production will require some more tweaking):

```
docker run -ti -p 4180:4180 --link <healthcheck_container_name>:healthchecks quay.io/oauth2-proxy/oauth2-proxy --cookie-secret=<set your secret> --client-id=<cognito app client id> --client-secret=<cognito app client secret> --cookie-secure=false --email-domain='*' --provider=oidc --redirect-url=http://localhost:4180/oauth2/callback --upstream=http://healthchecks:8000/ --oidc-issuer-url=https://cognito-idp.<your aws region>.amazonaws.com/<cognito pool id> --http-address=http://0.0.0.0:4180 --scope='email openid' --oidc-groups-claim='cognito:groups' --skip-auth-route='POST=^\/api\/.*,GET=^\/ping\/.*' --skip-provider-button=true --cookie-expire=1h --session-cookie-minimal=true
```

And set for healthchecks:
```
REMOTE_USER_HEADER=HTTP_X_FORWARDED_EMAIL
```
